### ISSUE TITLE
If the calendar type is selected in Windows AD, then the date picker shell will also be AD, even if the culture info is set to fa-IR.

### ISSUE THREAD

**Issue by Kamalifar** [bitfoundation/bitplatform#2188]
**Describe the bug**
I set the culture info in startup.configure as follows:

           `var cultureInfo = new CultureInfo("fa-IR");
            CultureInfo.DefaultThreadCurrentCulture = cultureInfo;
            CultureInfo.DefaultThreadCurrentUICulture = cultureInfo;
            CultureInfo.CurrentCulture = cultureInfo;
            CultureInfo.CurrentUICulture= cultureInfo;`

However, the date picker shell will also be AD as follows:

![Untitled](https://user-images.githubusercontent.com/35212420/173212872-ef7cbb1d-ec85-4ef8-b4da-c899267dcfe5.jpg)

The problem still persists even if the culture is set as follows in date picker:

`                    <BitDatePicker @bind-Value="LetterModel.AtfPyroDate" Style="width: 300px" AllowTextInput="true" FormatDate="yyyy/MM/dd" GoToToday="امروز" Culture="@(new System.Globalization.CultureInfo("fa-IR"))"></BitDatePicker>
`
Now, if the Windows calendar is set to Hijri, this problem will be solved as follows:
![image](https://user-images.githubusercontent.com/35212420/173213147-e519784d-8a2a-4036-8abf-47e0dee1a3dd.png)

The problem is that in Windows 7 or other operating systems, it may not be possible to set the calendar to Hijri.
How should we set the culture info so that the date picker shell is displayed in Persian in any case?

Expected behavior
It is better not to follow the system culture by setting the date picker culture anymore.

Bit Blazor client

Bit nuget package's version. 11.0.0
Blazor Server or Blazor Web Assembly or both? Blazor Server
Which browser? Chrome Version 102.0.5005.63 (Official Build) (64-bit)
Which blazor version? blazor 6

**Comment by ysmoradi** [comment]
I appreciate your detailed issue @Kamalifar
By Windows 7, is aspnetcore server that runs blazor server is running in Windows 7? Or only the browser is running in Windows 7 and server is somewhere else?

**Comment by Kamalifar** [comment]
Thank you for your prompt reply @ysmoradi.
Only the browser is running in Windows 7 and the server is somewhere else.
However, For different modes. Suppose the server operating system is a CentOS 5 and users are using Windows 7 or even Ubuntu.

**Comment by ysmoradi** [comment]
Ok, could you please provide a small repo that reproduces this?

> I also tried to customize the names of the months as shown below, but this time I did not succeed.
> `var cultureInfo = CultureInfo.CreateSpecificCulture("fa-IR"); cultureInfo.DateTimeFormat.MonthNames = new[] { "فروردین", "اردیبهشت", "خرداد", "تیر", "مرداد", "شهریور", "مهر", "آبان", "آذر", "دی", "بهمن", "اسفند", "" };`

Note that we're using this line of code in BitDatePicker
```cs
private CultureInfo culture = CultureInfo.CurrentUICulture;
```

**Comment by Kamalifar** [comment]
Unfortunately, I did not understand what I have to provide?

**Comment by ysmoradi** [comment]
A sample source code that has this issue

**Comment by Kamalifar** [comment]
I used this code in the following ways.

1)
`<BitDatePicker @bind-Value="LetterModel.AtfPyroDate" Style="width: 300px" AllowTextInput="true" FormatDate="yyyy/MM/dd" GoToToday="امروز" Culture="@cultureInfo"></BitDatePicker>`

`@code
{
 protected override void OnAfterRender(bool firstRender)
    {
        base.OnAfterRender(firstRender);
        if (firstRender)
        {
            cultureInfo.DateTimeFormat.MonthNames = new[] {"فروردین", "اردیبهشت", "خرداد", "تیر", "مرداد", "شهریور", "مهر", "آبان", "آذر", "دی", "بهمن", "اسفند", ""};
        }
    }
}`

2)
`<BitDatePicker @bind-Value="LetterModel.AtfPyroDate" Style="width: 300px" AllowTextInput="true" FormatDate="yyyy/MM/dd" GoToToday="امروز"></BitDatePicker>`

In startup:
` public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            var cultureInfo = CultureInfo.CreateSpecificCulture("fa-IR");
            cultureInfo.DateTimeFormat.MonthNames = new[] { "فروردین", "اردیبهشت", "خرداد", "تیر", "مرداد", "شهریور", "مهر", "آبان", "آذر", "دی", "بهمن", "اسفند", "" };
            CultureInfo.DefaultThreadCurrentCulture = cultureInfo;
            CultureInfo.DefaultThreadCurrentUICulture = cultureInfo;
            CultureInfo.CurrentCulture = cultureInfo;
            CultureInfo.CurrentUICulture = cultureInfo;`


3)
`<BitDatePicker @bind-Value="LetterModel.AtfPyroDate" Style="width: 300px" AllowTextInput="true" FormatDate="yyyy/MM/dd" GoToToday="امروز" Culture="@cultureInfo"></BitDatePicker>`

`@code
{
public CultureInfo cultureInfo = CultureInfo.CreateSpecificCulture("fa-IR");
    protected override void OnInitialized()
    {
        base.OnInitialized();
        cultureInfo.DateTimeFormat.MonthNames = new[] {"فروردین", "اردیبهشت", "خرداد", "تیر", "مرداد", "شهریور", "مهر", "آبان", "آذر", "دی", "بهمن", "اسفند", ""};
    }
}`

In the first mode, nothing happens. But in the second and third modes, only the month changes, but the year and day remain the same(AD).

![Untitled](https://user-images.githubusercontent.com/35212420/173227558-f3823de4-fed2-4a12-b604-b0d353a14e48.png)

**Comment by msynk** [comment]
@Kamalifar 
could you please create a new GitHub repository, push your code into it, and post the address of that repo?

**Comment by msynk** [comment]
BTW, I think it's an issue of the OS itself since in windows 10 and 11 we are seeing such an experience

**Comment by Kamalifar** [comment]
> I think it's an issue of the OS itself since in windows 10 and 11 we are not seeing such an experience

Exactly, I am testing this on Windows 10. This problem occurs if the Gregorian calendar is selected, but if the Hijri calendar is selected, the problem will be solved.

**Comment by Kamalifar** [comment]
> you please create a new GitHub repository, push your code into it, and post the address of that repo?

https://github.com/Kamalifar/TestBitComponentDatePicker/

**Comment by maryamhdr** [comment]
> **Describe the bug** I set the culture info in startup.configure as follows:
> 
> ```cs
> var cultureInfo = new CultureInfo("fa-IR");
> CultureInfo.DefaultThreadCurrentCulture = cultureInfo;
> CultureInfo.DefaultThreadCurrentUICulture = cultureInfo;
> CultureInfo.CurrentCulture = cultureInfo;
> CultureInfo.CurrentUICulture= cultureInfo;
> ```
> 
> However, the date picker shell will also be AD as follows, Only the names of the Gregorian months are shown in Persian.:
> 
> ![Untitled](https://user-images.githubusercontent.com/35212420/173212872-ef7cbb1d-ec85-4ef8-b4da-c899267dcfe5.jpg)
> 
> The problem still persists even if the culture is set as follows in date picker:
> 
> ```
> <BitDatePicker @bind-Value="LetterModel.AtfPyroDate" Style="width: 300px" AllowTextInput="true" FormatDate="yyyy/MM/dd" GoToToday="امروز" Culture="@(new System.Globalization.CultureInfo("fa-IR"))"></BitDatePicker>
> ```
> 
> Now, if the Windows calendar is set to Hijri, this problem will be solved as follows:
> 
> ![Untitled](https://user-images.githubusercontent.com/35212420/173213308-924eb923-df28-4248-be4e-ff28a5e6d336.jpg)
> 
> I also tried to customize the names of the months as shown below, but this time I did not succeed.
> 
> ```cs
> var cultureInfo = CultureInfo.CreateSpecificCulture("fa-IR");
> cultureInfo.DateTimeFormat.MonthNames = new[] { "فروردین", "اردیبهشت", "خرداد", "تیر", "مرداد", "شهریور", "مهر", "آبان", "آذر", "دی", "بهمن", "اسفند", "" };
> ```
> 
> The problem is that in Windows 7 or other operating systems, it may not be possible to set the calendar to Hijri. How should we set the culture info so that the date picker shell is displayed in Persian in any case?
> 
> Expected behavior It is better not to follow the system culture by setting the date picker culture anymore.
> 
> Bit Blazor client
> 
> Bit nuget package's version. 11.0.0 Blazor Server or Blazor Web Assembly or both? Blazor Server Which browser? Chrome Version 102.0.5005.63 (Official Build) (64-bit) Which blazor version? blazor 6

@Kamalifar 
Thanks a lot for contributing and submitting such a detailed and clear issue.
If you want to use the `BitPlatform` products widely in your projects, we can give you priority support for free, so in case you have an issue, you can contact the `BitPlatform` development team at MicrosiftTeams very quickly and get your answer ASAP.
If you would like to have this support plan, let us know and send us your email.

**Comment by Kamalifar** [comment]
Dear @maryamhdr,

Thanks for your support and explanation.
I'd like to use the BitDatePicker widely in my projects, So my email address is kamalifar@chmail.ir.

Best Regards,
Ali Kamalifar

**Comment by ysmoradi** [comment]
You've been invited to the Microsoft teams. Give it a try! @Kamalifar

**Comment by ysmoradi** [comment]
I was able to achieve followings

![image](https://user-images.githubusercontent.com/5352980/177537154-8865c8aa-b832-4b81-ad01-7c812fb8afa2.png)

Source code:

https://github.com/bitfoundation/bit-components-playground/tree/issue-2188

There is an issue with CultureInfo.TextInfo.IsRightToLeft property which needs more investigation

**Comment by Kamalifar** [comment]
Thank you so much for all you have done.
I used your code and the result is the same as before, as you can see in the picture.
![Untitled](https://user-images.githubusercontent.com/35212420/178089784-2218c9f0-c9e0-4cc8-894d-b43f43084b1e.png)

Source code:
https://github.com/Kamalifar/TestBitComponentDatePicker/blob/master/Pages/TestCultureInfoHelper.razor

Although today's date is selected, which means the 9th day of July, it displays the calendar of Mehr 2022, which means that only the month array mapping is done and the main date picker of the Gregorian calendar remains. July is the seventh month of the year and Mehr is the seventh month of the Shamsi year.
Please, first change the server calendar to Gregorian and then run the project. If the server calendar is changed while running, then the problem will not be seen.

Regards.

**Comment by ysmoradi** [comment]
Please use latest version of "Bit.BlazorUI" package, not "Bit.Client.Web.BlazorUI".
And could you please let me know if https://github.com/bitfoundation/bit-components-playground/tree/master/src/client/Pages/Issues/2188
Works for you by cloning and running that repo
Thanks in advance

**Comment by Kamalifar** [comment]
Thank you and congratulations.
https://github.com/bitfoundation/bit-components-playground/tree/master/src/client/Pages/Issues/2188 works for me very well, but it only works for WebAssembly, and unfortunately it doesn't work for BlazorServer.
It seems that the reflection has been useful, but probably some changes should be made for BlazorServer.

I updated https://github.com/Kamalifar/TestBitComponentDatePicker based on "Bit.BlazorUI" package, however, the result is the same as before.

**Comment by ysmoradi** [comment]
I ran your sample and it works like a chart. It seems that there is something OS-dependent is causing the issue
Could you please run your sample in Clean Windows 10 in Virtual Machine or elsewhere?
We can test Windows Server / Linux / Windows 7 after that.
Plus provide me cultureData values in your current setup as followings:
![image](https://user-images.githubusercontent.com/5352980/178103827-2b967a44-b420-4919-804b-7084feaba1ad.png)
This is your sample repo working totally fine on Windows 10
It seems that you've made it better by using fonts (Looks better than my sample)
![image](https://user-images.githubusercontent.com/5352980/178103874-88638b13-10e4-40a6-a88c-3c67530616d0.png)

**Comment by ysmoradi** [comment]
It has some differences. Your default calendar is GreorianCalendar, mine is PersianCalendar
That's the issue

**Comment by Kamalifar** [comment]
Yes, exactly, but why is it still affected by the operating system calendar? only in Blazor Server.

The below picture is my cultureData value in WebAssembly test project.
![Untitled](https://user-images.githubusercontent.com/35212420/178206103-b84dd473-2cdf-4513-924f-e44be712140f.png)

**Comment by ysmoradi** [comment]
Try this approach and let un know if this works or not
https://stackoverflow.com/a/15826881/2720104

**Comment by Kamalifar** [comment]
I tried to do this, but it can't do the cast on the following line.
```cs
newOptionalCalendarIDs[i] = (Int32) idProperty.GetValue(optionalCalendars[i], null);
````
I will leave the [address of the code](https://github.com/Kamalifar/TestBitComponentDatePicker/blob/master/Pages/PersianCulture.cs), please do a test.

**Comment by ysmoradi** [comment]
I'm afraid we can't truly test this because this needs an OS which problem exists in there and we don't have such an OS installed. And believe me or not, this is not BitDatePicker's issue, but we'll merge any PR you make if you wanna put more effort into this

**Comment by ysmoradi** [comment]
We're going to use `CultureInfo.DateTimeFormat.Calendar` instead of `CultureInfo.Calendar`
This property is writable and you can change that to `PersianCalendar` easily

**Comment by msynk** [comment]
The issue is accepted and planned. Resolving it will start ASAP.

**Comment by msynk** [comment]
Resolving this issue has started. It will be announced when this issue is resolved.

**Comment by msynk** [comment]
The issue has been resolved.
Note that in this state, the issue itself gets closed.

**Comment by Kamalifar** [comment]
> We're going to use `CultureInfo.DateTimeFormat.Calendar` instead of `CultureInfo.Calendar` This property is writable and you can change that to `PersianCalendar` easily

You're right, however, I do this too but still the calendar stays in Gregorian mode.
![Untitled](https://user-images.githubusercontent.com/35212420/178643501-e8b45f64-9383-43e3-96a7-f8a273c90700.png)

**Comment by ysmoradi** [comment]
In the new pre-release we're using `CultureInfo.DateTimeInfo.Calendar` which is writable
No need to change the default calendar anymore.

**Comment by Kamalifar** [comment]
Dear @msynk @ysmoradi @maryamhdr,
Congratulations!
The problem has been resolved definitely.
Thank you so much,

Best regards.
